### PR TITLE
feat: custom policy + hardened trust relationship

### PR DESCRIPTION
**Pull request by bt-macole**
## what

Updated trust policy so only the codebuild project in this module can use the iam role.
Added ability to toggle on/off default permissions
Added support for attaching a custom policy

> Fully backward compatible 
## why

fe75886860d4318705bce5a466df8a579fb5f85f

    adding a condition on the trust policy for the codebuild project arn ensures
    the iam role can not be used by any other codebuild project.
    
    codebuild projects could be extremely permissive, even when least privileged
    so enforcing the role can only be used by the intended codebuild project
    limits to ability for a threat actor to quietly take control of a powerful
    role and do threat actory things.

9f0163f70f0239d00e28a4801ab3cbde8cd8fe6d

    the default permissions are good for getting up and running, however, they are far more
    permissive than any least privileged policy would like, being `*` for all resources for
    anything in the default list + any and all additional_permissions.
    
    this allows users to still utilize the quick up and running policy, while also being able
    to disable it and replace it with a least privileged custom policy.
    
    the lifecycle rule ensure that users don't get confused by additional permissions and custom
    policy variables, failing on a plan if they attempt to use additional_permissions with
    default_permissions_enabled set to `false`.

### Test Output:

```terraform
    --- PASS: TestExamplesCustom (62.77s)
    --- PASS: TestExamplesComplete (62.87s)
    --- PASS: TestExamplesVPC (78.32s)
    PASS
    ok      github.com/cloudposse/terraform-aws-codebuild   78.790s
```

## references

resolves: #131

### CODE REVIEW

**Comment by bt-macole** [comment]
@johncblandii or @srhopkins looks like you were assigned this PR. Is there anything else you need from me to get this in?

we'd like to take advantage of this module, but need these changes for security reasons (https://github.com/cloudposse/terraform-aws-codebuild/issues/131)

File: variables.tf [hunk 1]
```diff
@@ -254,10 +254,22 @@ variable "logs_config" {
   description = "Configuration for the builds to store log data to CloudWatch or S3."
 }
 
+variable "default_permissions_enabled" {
```
Comment (joe-niland) [inline variables.tf:257]: Could we use the presence of the custom policy instead of this flag?

File: variables.tf [hunk 2]
```diff
@@ -254,10 +254,22 @@ variable "logs_config" {
...
+  type        = list(string)
   default     = []
-  description = "List of action strings which will be added to IAM service account permissions."
+  description = "List of action strings which will be added to IAM service account permissions. Only used if `default_permissions_enabled` is set to true."
+}
+
+variable "custom_policy" {
+  type        = string
+  description = "Jsonencoded IAM policy to add to the IAM service account permissions."
+  default     = ""
```
Comment (joe-niland) [inline variables.tf:272]: For optional strings like this, the Cloud Posse convention is to use a list of string (in this case), with the default being `[]`. We've written this up but the [PR is still open](https://github.com/cloudposse/.github/pull/30/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R133).

File: variables.tf [hunk 3]
```diff
@@ -254,10 +254,22 @@ variable "logs_config" {
...
-  type        = list(any)
+  type        = list(string)
   default     = []
-  description = "List of action strings which will be added to IAM service account permissions."
+  description = "List of action strings which will be added to IAM service account permissions. Only used if `default_permissions_enabled` is set to true."
+}
+
+variable "custom_policy" {
+  type        = string
+  description = "Jsonencoded IAM policy to add to the IAM service account permissions."
```
Comment (joe-niland) [inline variables.tf:271]:
```suggestion
  description = "JSON encoded IAM policy to add to the IAM service account permissions."
```

**Review by joe-niland — COMMENTED** [review]
Thanks @bt-macole - this is a really good improvement.
A couple of comments.

File: variables.tf (hunk 1 above)
Comment (bt-macole) [reply variables.tf:257]: it's possible one wants the defaults and custom policy. I used the enable variable to follow conventions elsewhere in the module.

happy to change that, if you'd prefer it to be an either or.

File: variables.tf (hunk 2 above)
Comment (bt-macole) [reply variables.tf:272]: sure thing, that should play nicely with how this is merged together with the other policies as well.

File: variables.tf [hunk 4]
```diff
@@ -254,10 +254,27 @@ variable "logs_config" {
   description = "Configuration for the builds to store log data to CloudWatch or S3."
 }
 
+variable "default_permissions_enabled" {
+  type        = bool
+  default     = true
+  description = "When 'true' default base iam permissions to get up and running with codebuild are attached. Those who want a least priveleged policy can instead set to `false` and use the `custom policy` variable."
```
Comment (joe-niland) [inline variables.tf:260]:
```suggestion
  description = "When 'true' default base IAM permissions to get up and running with CodeBuild are attached. Those who want a least privileged policy can instead set to `false` and use the `custom_policy` variable."
```

File: examples/custom/variables.tf [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  description = "The type of storage that will be used for the AWS CodeBuild project cache. Valid values: NO_CACHE, LOCAL, and S3.  Defaults to NO_CACHE.  If cache_type is S3, it will create an S3 bucket for storing codebuild cache inside"
+}
+
+variable "default_permissions_enabled" {
+  type        = bool
+  default     = true
+  description = "When 'true' default base iam permissions to get up and running with codebuild are attached. Those who want a least priveleged policy can instead set to `false` and use the `custom policy` variable."
+}
+
+variable "custom_policy" {
```
Comment (joe-niland) [inline examples/custom/variables.tf:45]: This one would need updating to match the variable change

File: examples/custom/main.tf [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  source                      = "../../"
+  description                 = "This is my awesome CodeBuild project"
+  concurrent_build_limit      = 10
+  cache_bucket_suffix_enabled = var.cache_bucket_suffix_enabled
+  environment_variables       = var.environment_variables
+  cache_expiration_days       = var.cache_expiration_days
+  cache_type                  = var.cache_type
+
+  default_permissions_enabled = var.default_permissions_enabled
+  custom_policy               = [jsonencode(var.custom_policy)]
```
Comment (joe-niland) [inline examples/custom/main.tf:15]: Update to hande list type

**Review by joe-niland — COMMENTED** [review]
Thanks @bt-macole - a couple more things

File: examples/custom/variables.tf (hunk 5 above)
Comment (bt-macole) [reply examples/custom/variables.tf:45]: updated and reran tests

```terraform
--- PASS: TestExamplesComplete (61.73s)
--- PASS: TestExamplesCustom (62.47s)
--- PASS: TestExamplesVPC (87.75s)
PASS
ok      github.com/cloudposse/terraform-aws-codebuild   88.190s
```

**Comment by joe-niland** [comment]
I think you'll need to re-run 

```
terraform fmt
make init
make readme
```

**Comment by bt-macole** [comment]
> I think you'll need to re-run
> 
> ```
> terraform fmt
> make init
> make readme
> ```

hehe my bad! too used to precommit running the terraform fmt and terraform-docs in our repo 🤦

**Review by joe-niland — APPROVED** [review]

**Review by joe-niland — APPROVED** [review]

**Comment by joe-niland** [comment]
Thanks @bt-macole we got there!
Released as https://github.com/cloudposse/terraform-aws-codebuild/releases/tag/2.0.2

**Pull request merged**